src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,164 @@
+     *  * contentType (string): DEPRECATED content type to be stored with the file.
+     *    This information should now be added to the metadata
+     *
+     *  * aliases (array of strings): DEPRECATED An array of aliases.
+     *    Applications wishing to store aliases should add an aliases field to the
+     *    metadata document instead.
+     *
+     *  * metadata (array or object): User data for the 'metadata' field of the files
+     *    collection document.
+     *
+     *  * writeConcern (MongoDB\Driver\WriteConcern): Write concern.
+     *
+     * @param \MongoDB\Collection    $filesCollection  Files Collection
+     * @param \MongoDB\Collection    $chunksCollection Chunks Collection
+     * @param int32                  $chunkSizeBytes   Size of chunk
+     * @param string                 $filename         Filename to insert
+     * @param array                  $options          File options
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        \MongoDB\Collection $filesCollection,
+        \MongoDB\Collection $chunksCollection,
+        $chunkSizeBytes,
+        $filename,
+        array $options=[]
+        )
+    {
+        $this->ctx = hash_init('md5');
+        $uploadDate = time();
+        $objectId = new \MongoDB\BSON\ObjectId();
+        $main_file = [
+            "chunkSize" => $chunkSizeBytes,
+            "filename" => $filename,
+            "uploadDate" => $uploadDate,
+            "_id" => $objectId
+        ];
+
+        $fileOptions = [];
+        if (isset($options['contentType'])) {
+            if (is_string($options['contentType'])) {
+                $fileOptions['contentType'] = $options['contentType'];
+            } else {
+                throw new InvalidArgumentTypeException('"contentType" option', $options['contentType'], 'string');
+            }
+        }
+        if (isset($options['aliases'])) {
+            if (\MongoDB\is_string_array($options['aliases'])) {
+                $fileOptions['aliases'] = $options['aliases'];
+            } else {
+                throw new InvalidArgumentTypeException('"aliases" option', $options['aliases'], 'array of strings');
+            }
+        }
+
+        if (isset($options['metadata'])) {
+            if (is_array($options['metadata']) || is_object($options['metadata'])) {
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($filesCollection, $chunksCollection, $chunkSizeBytes);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        while ($data = fread($source, $this->chunkSizeBytes)) {
```

Comment:
2 Questions on this:

- 1 What do you mean by 'should also cover the Bucket class' `uploadFromStream()` method? ' Like it will be api compliant ? 
- 2 I'm not totally sure what errors should be being thrown if these checks don't succeed.